pkg/cloud/gcp/gcs_kms.go

```diff
@@ -0,0 +1,196 @@
+// licenses/APL.txt.
+
+package gcp
+
+import (
+	"context"
+	"encoding/base64"
+	"hash/crc32"
+	"net/url"
+	"strings"
+
+	kms "cloud.google.com/go/kms/apiv1"
+	"github.com/cockroachdb/cockroach/pkg/cloud"
+	"github.com/cockroachdb/errors"
+	"google.golang.org/api/option"
+	kmspb "google.golang.org/genproto/googleapis/cloud/kms/v1"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+)
+
+const gcsScheme = "gs"
+
+type gcsKMS struct {
+	kms                 *kms.KeyManagementClient
+	customerMasterKeyID string
+}
+
+var _ cloud.KMS = &gcsKMS{}
+
+func init() {
+	cloud.RegisterKMSFromURIFactory(MakeGCSKMS, gcsScheme)
+}
+
+type kmsURIParams struct {
+	credentials string
+	auth        string
+}
+
+func resolveKMSURIParams(kmsURI url.URL) kmsURIParams {
+	params := kmsURIParams{
+		credentials: kmsURI.Query().Get(CredentialsParam),
+		auth:        kmsURI.Query().Get(cloud.AuthParam),
+	}
+
+	return params
+}
+
+// MakeGCSKMS is the factory method which returns a configured, ready-to-use
+// GCS KMS object.
+func MakeGCSKMS(uri string, env cloud.KMSEnv) (cloud.KMS, error) {
+	if env.KMSConfig().DisableOutbound {
+		return nil, errors.New("external IO must be enabled to use GCS KMS")
+	}
+	kmsURI, err := url.ParseRequestURI(uri)
+	if err != nil {
+		return nil, err
+	}
+
+	// Extract the URI parameters required to setup the GCS KMS session.
+	kmsURIParams := resolveKMSURIParams(*kmsURI)
+
+	// Client options to authenticate and start a GCS KMS session.
+	// Currently only accepting json of service account.
+	var credentialsOpt []option.ClientOption
+
+	switch kmsURIParams.auth {
+	case "", cloud.AuthParamSpecified:
+		if kmsURIParams.credentials == "" {
+			return nil, errors.Errorf(
+				"%s is set to '%s', but %s is not set",
+				cloud.AuthParam,
+				cloud.AuthParamSpecified,
+				CredentialsParam,
+			)
+		}
+
+		// Credentials are passed in base64 encoded, so decode the credentials first.
+		credentialsJSON, err := base64.StdEncoding.DecodeString(kmsURIParams.credentials)
+		if err != nil {
+			return nil, err
+		}
+
+		credentialsOpt = append(credentialsOpt, option.WithCredentialsJSON(credentialsJSON))
+	case cloud.AuthParamImplicit:
+		if env.KMSConfig().DisableImplicitCredentials {
+			return nil, errors.New(
+				"implicit credentials disallowed for gcs due to --external-io-implicit-credentials flag")
+		}
+		// If implicit credentials used, no client options needed.
+	default:
+		return nil, errors.Errorf("unsupported value %s for %s", kmsURIParams.auth, cloud.AuthParam)
+	}
+
+	ctx := context.Background()
+
+	kmc, err := kms.NewKeyManagementClient(ctx, credentialsOpt...)
+
+	if err != nil {
+		return nil, err
+	}
+
+	// Check the given cmk to see if key version is given.
+	// If it is, strip it from the cmk.
+	cmkID := strings.TrimPrefix(kmsURI.Path, "/")
+	versionIndex := strings.Index(cmkID, "/cryptoKeyVersions/")
+	if versionIndex != -1 {
+		cmkID = strings.Split(cmkID, "/cryptoKeyVersions/")[0]
+	}
+
```

Comment:
nit: can we collapse into:
```
cmkID = strings.Split(cmkID, "/crypto...")[0]
```

I believe if the separator doesnt exist in the string split returns a single element slice with the string.